

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Sequence — Extension of Bio.Align &mdash; epitopsy 0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="../_static/copybutton.js"></script>
    <link rel="top" title="epitopsy 0.1 documentation" href="../index.html" />
    <link rel="up" title="Epitopsy Package" href="../epitopsy.html" />
    <link rel="next" title="Structure — Additional classes for PDB files" href="Structure.html" />
    <link rel="prev" title="DXFile — YYYYYYY" href="DXFile.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Structure.html" title="Structure — Additional classes for PDB files"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="DXFile.html" title="DXFile — YYYYYYY"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">epitopsy 0.1 documentation</a> &raquo;</li>
          <li><a href="../epitopsy.html" accesskey="U">Epitopsy Package</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="module-Sequence">
<span id="sequence-extension-of-bio-align"></span><h1><a class="reference internal" href="#module-Sequence" title="Sequence: Shortcuts for the automation of repetitive tasks involving biopython alignment objects."><tt class="xref py py-mod docutils literal"><span class="pre">Sequence</span></tt></a> &#8212; Extension of <cite>Bio.Align</cite><a class="headerlink" href="#module-Sequence" title="Permalink to this headline">¶</a></h1>
<p>This module provides various shortcuts for the automation of repetitive tasks involving biopython alignment objects.</p>
<div class="section" id="module-syntax">
<span id="sequence-syntax"></span><h2>Module Syntax<a class="headerlink" href="#module-syntax" title="Permalink to this headline">¶</a></h2>
<p>The tools in this package generally take a <tt class="xref py py-class docutils literal"><span class="pre">Bio.Align.MultipleSeqAlignment</span></tt> object as input and return a list which can be used for the instanciation of a new <tt class="xref py py-class docutils literal"><span class="pre">Bio.Align.MultipleSeqAlignment</span></tt> object.</p>
</div>
<div class="section" id="module-contents">
<span id="contents-of-module-sequence"></span><h2>Module Contents<a class="headerlink" href="#module-contents" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="Sequence.make_alignment_upper_case">
<tt class="descclassname">Sequence.</tt><tt class="descname">make_alignment_upper_case</tt><big>(</big><em>orig_aln</em><big>)</big><a class="headerlink" href="#Sequence.make_alignment_upper_case" title="Permalink to this definition">¶</a></dt>
<dd><p>Transform all characters to upper case.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>orig_aln</strong> (<em>obj</em>) &#8211; biopython alignment, instance of <tt class="xref py py-class docutils literal"><span class="pre">Bio.Align.MultipleSeqAlignment</span></tt></td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">upper case biopython alignment, feed for <tt class="xref py py-class docutils literal"><span class="pre">Bio.Align.MultipleSeqAlignment</span></tt></td>
</tr>
</tbody>
</table>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s">&quot;/path/seq.txt&quot;</span><span class="p">,</span> <span class="s">&quot;rU&quot;</span><span class="p">),</span> <span class="s">&quot;fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">alignment</span>
<span class="go">SingleLetterAlphabet() alignment with 3 rows and 178 columns</span>
<span class="go">iigp--gr-gfgkrrhpkkltplaykqfipnvaektlgasgrye...sgg 3M1N:A|PDBID|CHAIN|SEQUENCE</span>
<span class="go">iigp--grpgfgkrrhpkkltplaykqfipnvaektlgasgrye...sgg PRO1:A|NAME1|CHAIN|SEQUENCE</span>
<span class="go">iigpxxgrcgfgkrrhpkkltplaykqfipnvaektlgasgrye...sgg PRO2:A|NAME2|CHAIN|SEQUENCE</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">new_alignment</span> <span class="o">=</span> <span class="n">Sequence</span><span class="o">.</span><span class="n">make_alignment_upper_case</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">Align</span><span class="o">.</span><span class="n">MultipleSeqAlignment</span><span class="p">(</span><span class="n">new_alignment</span><span class="p">)</span>
<span class="go">Alphabet() alignment with 3 rows and 178 columns</span>
<span class="go">IIGP--GR-GFGKRRHPKKLTPLAYKQFIPNVAEKTLGASGRYE...SGG 3M1N:A|PDBID|CHAIN|SEQUENCE</span>
<span class="go">IIGP--GRPGFGKRRHPKKLTPLAYKQFIPNVAEKTLGASGRYE...SGG PRO1:A|NAME1|CHAIN|SEQUENCE</span>
<span class="go">IIGPXXGRCGFGKRRHPKKLTPLAYKQFIPNVAEKTLGASGRYE...SGG PRO2:A|NAME2|CHAIN|SEQUENCE</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="Sequence.make_alignment_clean">
<tt class="descclassname">Sequence.</tt><tt class="descname">make_alignment_clean</tt><big>(</big><em>orig_aln</em><big>)</big><a class="headerlink" href="#Sequence.make_alignment_clean" title="Permalink to this definition">¶</a></dt>
<dd><p>Deletes sequences with non standard amino acids and reduce the
alignment size if unnecessary gaps are found (this may happen
when a sequence is deleted).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>orig_aln</strong> (<em>obj</em>) &#8211; biopython alignment, instance of <tt class="xref py py-class docutils literal"><span class="pre">Bio.Align.MultipleSeqAlignment</span></tt></td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">a clean biopython alignment, feed for <tt class="xref py py-class docutils literal"><span class="pre">Bio.Align.MultipleSeqAlignment</span></tt></td>
</tr>
</tbody>
</table>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s">&quot;/path/seq.txt&quot;</span><span class="p">,</span> <span class="s">&quot;rU&quot;</span><span class="p">),</span> <span class="s">&quot;fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">alignment</span>
<span class="go">SingleLetterAlphabet() alignment with 3 rows and 178 columns</span>
<span class="go">IIGP--GR-GFGKRRHPKKLTPLAYKQFIPNVAEKTLGASGRYE...SGG 3M1N:A|PDBID|CHAIN|SEQUENCE</span>
<span class="go">IIGP--GRPGFGKRRHPKKLTPLAYKQFIPNVAEKTLGASGRYE...SGG PRO1:A|NAME1|CHAIN|SEQUENCE</span>
<span class="go">IIGPXXGRCGFGKRRHPKKLTPLAYKQFIPNVAEKTLGASGRYE...SGG PRO2:A|NAME2|CHAIN|SEQUENCE</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">clean_alignment</span> <span class="o">=</span> <span class="n">Sequence</span><span class="o">.</span><span class="n">make_alignment_clean</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="go">input size: 3 x 178 (#seq x len(seq))</span>
<span class="go">output size: 2 x 176</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">Align</span><span class="o">.</span><span class="n">MultipleSeqAlignment</span><span class="p">(</span><span class="n">clean_alignment</span><span class="p">)</span>
<span class="go">Alphabet() alignment with 2 rows and 176 columns</span>
<span class="go">IIGPGR-GFGKRRHPKKLTPLAYKQFIPNVAEKTLGASGRYEGK...SGG 3M1N:A|PDBID|CHAIN|SEQUENCE</span>
<span class="go">IIGPGRPGFGKRRHPKKLTPLAYKQFIPNVAEKTLGASGRYEGK...SGG PRO1:A|NAME1|CHAIN|SEQUENCE</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># 3rd sequence with &#39;X&#39; was deleted; two &#39;-&#39; became useless and were removed</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="Sequence.format_stockholm_to_fasta">
<tt class="descclassname">Sequence.</tt><tt class="descname">format_stockholm_to_fasta</tt><big>(</big><em>input_path</em>, <em>output_path</em><big>)</big><a class="headerlink" href="#Sequence.format_stockholm_to_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Open a stockholm formatted alignment file located at <em>input_path</em> and
write the corresponding fasta formatted alignment to <em>output_path</em>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>input_path</strong> (<em>str</em>) &#8211; path to the stockholm formatted alignment</li>
<li><strong>output_path</strong> (<em>str</em>) &#8211; path to the new fasta formatted alignment</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last"><tt class="docutils literal"><span class="pre">None</span></tt></p>
</td>
</tr>
</tbody>
</table>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">Sequence</span><span class="o">.</span><span class="n">format_stockholm_to_fasta</span><span class="p">(</span><span class="s">&quot;/path/stock.txt&quot;</span><span class="p">,</span> <span class="s">&quot;/path/fasta.txt&quot;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="Sequence.get_pairwise_alignment">
<tt class="descclassname">Sequence.</tt><tt class="descname">get_pairwise_alignment</tt><big>(</big><em>ref_seq</em>, <em>query_seq</em><big>)</big><a class="headerlink" href="#Sequence.get_pairwise_alignment" title="Permalink to this definition">¶</a></dt>
<dd><p>Match the query sequence <em>query_seq</em> with the reference sequence <em>ref_seq</em>
and return the corresponding  aligned sequences in a dictionary with keys
(&#8220;ref_seq&#8221;, &#8220;query_seq&#8221;). It uses the first calculated alignment, equally
scored alignments are not considered.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>ref_seq</strong> (<em>str</em>) &#8211; string of the reference sequence</li>
<li><strong>query_seq</strong> (<em>str</em>) &#8211; string of the query sequence</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">a dictionary with keys (&#8220;ref_seq&#8221;, &#8220;query_seq&#8221;)</p>
</td>
</tr>
</tbody>
</table>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">align</span> <span class="o">=</span> <span class="n">Sequence</span><span class="o">.</span><span class="n">get_pairwise_alignment</span><span class="p">(</span><span class="s">&quot;IIGPTTGRGFGKRR&quot;</span><span class="p">,</span><span class="s">&quot;IIGPGGGRGFGKRR&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">seq</span> <span class="ow">in</span> <span class="n">align</span><span class="o">.</span><span class="n">values</span><span class="p">():</span> <span class="k">print</span> <span class="n">seq</span>
<span class="gp">...</span>
<span class="go">IIGPTTG--RGFGKRR</span>
<span class="go">IIGP--GGGRGFGKRR</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="Sequence.get_almost_conserved_columns">
<tt class="descclassname">Sequence.</tt><tt class="descname">get_almost_conserved_columns</tt><big>(</big><em>aln</em>, <em>cutoff=0.95</em><big>)</big><a class="headerlink" href="#Sequence.get_almost_conserved_columns" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of the positions in the aligned sequence where the amino
acid with the highest occurence has a frequency larger than the cutoff
(&#8216;-&#8216; is also counted as an amino acid). The column index starts with 1!</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>aln</strong> (<em>obj</em>) &#8211; biopython alignment, instance of <tt class="xref py py-class docutils literal"><span class="pre">Bio.Align.MultipleSeqAlignment</span></tt></li>
<li><strong>cutoff</strong> (<em>float</em>) &#8211; minimal frequency of occurence</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">list with the columns index</p>
</td>
</tr>
</tbody>
</table>
<p>Example:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">alignment</span> <span class="o">=</span> <span class="n">AlignIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="nb">open</span><span class="p">(</span><span class="s">&quot;/path/fasta.txt&quot;</span><span class="p">,</span> <span class="s">&quot;rU&quot;</span><span class="p">),</span> <span class="s">&quot;fasta&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">c</span> <span class="o">=</span> <span class="n">Sequence</span><span class="o">.</span><span class="n">get_almost_conserved_columns</span><span class="p">(</span><span class="n">alignment</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">c</span>
<span class="go">[1, 3, 4, 7, 8, 10, 11, 12, 14, 16]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c"># Result visualization</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">print</span> <span class="n">alignment</span><span class="p">;</span> <span class="k">print</span> <span class="s">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s">&#39;+&#39;</span> <span class="k">if</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">c</span> <span class="k">else</span> <span class="s">&#39;-&#39;</span> <span class="k">for</span> <span class="n">x</span>
<span class="gp">... </span><span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="n">alignment</span><span class="o">.</span><span class="n">get_alignment_length</span><span class="p">()</span><span class="o">+</span><span class="mi">1</span><span class="p">))</span> <span class="o">+</span> <span class="s">&quot; (at 95+%)&quot;</span>
<span class="go">SingleLetterAlphabet() alignment with 3 rows and 16 columns</span>
<span class="go">IKGP--GR-GFGPRRH prot1</span>
<span class="go">IIGPRGGRPGFGKRVH prot2</span>
<span class="go">IIGPGGGRCGFGKRVH prot3</span>
<span class="go">+-++--++-+++-+-+ (at 95+%)</span>
</pre></div>
</div>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#"><tt class="docutils literal"><span class="pre">Sequence</span></tt> &#8212; Extension of <cite>Bio.Align</cite></a><ul>
<li><a class="reference internal" href="#module-syntax">Module Syntax</a></li>
<li><a class="reference internal" href="#module-contents">Module Contents</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="DXFile.html"
                        title="previous chapter"><tt class="docutils literal"><span class="pre">DXFile</span></tt> &#8212; YYYYYYY</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="Structure.html"
                        title="next chapter"><tt class="docutils literal"><span class="pre">Structure</span></tt> &#8212; Additional classes for PDB files</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/packages/Sequence.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Structure.html" title="Structure — Additional classes for PDB files"
             >next</a> |</li>
        <li class="right" >
          <a href="DXFile.html" title="DXFile — YYYYYYY"
             >previous</a> |</li>
        <li><a href="../index.html">epitopsy 0.1 documentation</a> &raquo;</li>
          <li><a href="../epitopsy.html" >Epitopsy Package</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013-2014, Christoph Wilms.
      Last updated on Jan 08, 2014.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>